src/devices/src/virtio/balloon/device.rs

```diff
@@ -0,0 +1,920 @@
+            error!("Failed to get queue event: {:?}", e);
+            METRICS.balloon.event_fails.inc();
+        } else if self.process_deflate_queue() {
+            self.signal_used_queue()
+                .unwrap_or_else(report_balloon_event_fail);
+        }
+    }
+
+    pub(crate) fn process_stats_queue_event(&mut self) {
+        if let Err(e) = self.queue_evts[STATS_INDEX].read() {
+            error!("Failed to get queue event: {:?}", e);
+            METRICS.balloon.event_fails.inc();
+        } else {
+            self.process_stats_queue()
+                .unwrap_or_else(report_balloon_event_fail);
+        }
+    }
+
+    pub(crate) fn process_stats_timer_event(&mut self) {
+        let mem = match self.device_state {
+            DeviceState::Activated(ref mem) => mem,
+            // This should never happen, it's been already validated in the event handler.
+            DeviceState::Inactive => unreachable!(),
+        };
+        self.stats_timer.read();
+
+        // The communication is driven by the device by using the buffer
+        // and sending a used buffer notification
+        if let Some(index) = self.stats_desc_index.take() {
+            self.queues[STATS_INDEX].add_used(&mem, index, 0);
+            self.signal_used_queue()
+                .unwrap_or_else(report_balloon_event_fail);
+        }
+    }
+
+    pub(crate) fn process_inflate(&mut self) -> std::result::Result<(), DeviceError> {
+        let mem = match self.device_state {
+            DeviceState::Activated(ref mem) => mem,
+            // This should never happen, it's been already validated in the event handler.
+            DeviceState::Inactive => unreachable!(),
+        };
+        METRICS.balloon.inflate_count.inc();
+
+        let queue = &mut self.queues[INFLATE_INDEX];
+        let mut pages = Vec::with_capacity(MAX_PAGES_IN_DESC);
+        let mut needs_interrupt = false;
+
+        while let Some(head) = queue.pop(&mem) {
+            let len = head.len;
+            if !head.is_write_only() && len % SIZE_OF_U32 as u32 == 0 {
+                for index in (0..len).step_by(SIZE_OF_U32) {
+                    let addr = head
+                        .addr
+                        .checked_add(index as u64)
+                        .ok_or(DeviceError::MalformedDescriptor)?;
+
+                    let page_frame_number = mem
+                        .read_obj::<u32>(addr)
+                        .map_err(|_| DeviceError::MalformedDescriptor)?;
+
+                    pages.push(page_frame_number);
+                }
+            }
+
+            // Acknowledge the receipt of the descriptor.
+            // 0 is number of bytes the device has written to memory.
+            queue.add_used(&mem, head.index, 0);
+            needs_interrupt = true;
+        }
+
+        if needs_interrupt {
+            self.signal_used_queue()
+                .unwrap_or_else(report_balloon_event_fail);
```

Comment:
Doesn't this lead to double counting of the event failure?

Reply:
No because if the `signal_used_queue()` call fails, the event will be logged as an error through `report_balloon_event_fail` and the execution of the function continues with the potential to return an `Ok` value, thus not reporting it again in `process_inflate_queue_event()`.